---
title: "Lab 10"
author: "Simin Manole"
date: "11/2/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Part I - Reading VCF data
### VCF file structure
```{r}
library(vcfR)
data(vcfR_example)
vcf
```

### The meta region
```{r}
strwrap(vcf@meta[1:7])
queryMETA(vcf)
queryMETA(vcf, element = 'DP')
queryMETA(vcf, element = 'FORMAT=<ID=DP')
```

### The fix region
```{r}
head(getFIX(vcf))
```

### The gt region
```{r}
vcf@gt[1:6, 1:4]
```

### vcfR
```{r}
vcf <- read.vcfR("data/pinfsc50_filtered.vcf.gz")
head(vcf)
write.vcf(vcf, "data/myVCFdata_filtered.vcf.gz")
```

## Exercises Part I
1) How would we find more information about read.vcfR()?
```{r, eval=FALSE}
?read.vcfR()
```


2) How would we learn what the acronym “AD” stands for?
```{r}
queryMETA(vcf, element = 'AD')
```


3) We used the head() function to view the first few lines of fix data. How would we view the last few lines of fix data?
```{r}
tail(vcf@fix)
```


4) There is a column in the fix portion of the data called QUAL. It is not defined in the meta portion of the data because it is defined in the VCF specification. It stands for ‘quality’. Does QUAL appear useful to us? Why or why not?

According to the plot, the VCF specification, and [the wikipedia article on the "Phred quality score"](https://en.wikipedia.org/wiki/Phred_quality_score), a score higher than 20 or 30 is acceptable data. The data appears to largely have an acceptable QUAL. 
```{r}
library(ggplot2)
qplot(getQUAL(vcf), geom = "histogram", log = "y", bins = 100)
```

5) How would we query the sample names?
"The gt (genotype) region contains information about each variant for each sample."
```{r}
colnames(vcf@gt)
```

## Part II - Analysis of Genome Data
### Opening and examining the dataset
```{r}
library('vcfR')
vcf <- read.vcfR("data/pinfsc50_filtered.vcf.gz")
```

### Converting VCF data to a genlight object
```{r}
library("adegenet")
x <- vcfR2genlight(vcf)
x
```

```{r}
# vcfR
gt <- extract.gt(vcf, element = "GT")
gt[c(2,6,18), 1:3]
```

```{r}
# genlight
t(as.matrix(x))[c(1,5,17), 1:3]
```

```{r}
pop(x) <- as.factor(c("us", "eu", "us", "af", "eu", "us", "mx", "eu", "eu", "sa", "mx", "sa", "us", "sa", "Pmir", "us", "eu", "eu"))
popNames(x)
```

```{r}
ploidy(x) <- 2
```

### Distance matrices
```{r}
x.dist <- dist(x)
library("poppr")
x.dist <- poppr::bitwise.dist(x)
```

### Creating chromR objects
```{r}
library(vcfR)

# Find the files.
vcf_file <- system.file("extdata", "pinf_sc50.vcf.gz", package = "pinfsc50")
dna_file <- system.file("extdata", "pinf_sc50.fasta", package = "pinfsc50")
gff_file <- system.file("extdata", "pinf_sc50.gff", package = "pinfsc50")

# Input the files.
vcf <- read.vcfR(vcf_file, verbose = FALSE)
dna <- ape::read.dna(dna_file, format = "fasta")
gff <- read.table(gff_file, sep="\t", quote="")

# Create a chromR object.
chrom <- create.chromR(name="Supercontig", vcf=vcf, seq=dna, ann=gff, verbose=TRUE)
```

```{r}
chrom
```

```{r}
plot(chrom)
```

```{r}
chromoqc(chrom, dp.alpha = 66)
```

## Processing chromR objects
```{r}
chrom <- proc.chromR(chrom, verbose = TRUE)
plot(chrom)
```

```{r}
chromoqc(chrom, dp.alpha = 66)
```

```{r}
#vcf <- read.vcfR("pinfsc50_qc.vcf.gz", verbose = FALSE)
vcf <- read.vcfR("data/pinfsc50_filtered.vcf.gz", verbose = FALSE)
chrom <- create.chromR(name="Supercontig", vcf=vcf, seq=dna, ann=gff, verbose=FALSE)
chrom <- proc.chromR(chrom, verbose = FALSE)
chromoqc(chrom, dp.alpha = 66)
```

### Tabular summaries
```{r}
head(chrom@var.info)
```

```{r}
head(chrom@win.info)
```

### Genetic differentiation
```{r}
library(vcfR)
data(vcfR_example)
pop <- as.factor(c("us", "eu", "us", "af", "eu", "us", "mx", "eu", "eu", "sa", "mx", "sa", "us", "sa", "Pmir", "us", "eu", "eu"))
myDiff <- genetic_diff(vcf, pops = pop, method = 'nei')
knitr::kable(head(myDiff[,1:15]))
```

```{r}
knitr::kable(head(myDiff[,16:19]))
```

```{r}
knitr::kable(round(colMeans(myDiff[,c(3:9,16,19)], na.rm = TRUE), digits = 3))
```

```{r}
library(reshape2)
library(ggplot2)

dpf <- melt(myDiff[,c(3:8,19)], varnames=c('Index', 'Sample'), value.name = 'Depth', na.rm=TRUE)
```

```{r}
p <- ggplot(dpf, aes(x=variable, y=Depth)) + geom_violin(fill="#2ca25f", adjust = 1.2)
p <- p + xlab("")
p <- p + ylab("")
p <- p + theme_bw()
p
```

## Exercises Part II

1) You actually have everything you need to make a Manhattan plot. Can you figure out how to plot G′ST (y-axis) by genomic position (POS)?
```{r}
plot(
  x = getPOS(vcf), 
  y = myDiff$Gprimest,  
  pch = 20, 
  col = "#00FF00", 
  xlab = "", 
  ylab = "", 
  ylim = c(0, 1), 
  xaxt = "n")

axis(side = 1, at = seq(0, 1e5, by = 1e4), labels = seq(0, 100, by = 10))
title(xlab='Genomic position (Kbp)')
title(ylab = expression(italic("G'"["ST"])))
```


2) This Manhatttan plot should look a bit unusual. Can you think of anything that may be wrong with this analysis?
```{r}
table(pop)
```


3) Can you figure out how to zoom in on a particular region of a chromosome in chromoqc()?
The x-axis is given a range to zoom in on area with DP.
```{r}
chromoqc(chrom, 
         dp.alpha = 66, 
         xlim = c(8e04, 7e05))
```


4) Can you use the function queryMETA() to look for other data in your file that may be of interest?
```{r}
queryMETA(vcf)
```

